### PR TITLE
document kube-apiserver identity

### PR DESCRIPTION
**Pull request by roycaihw**
enhancement issue: https://github.com/kubernetes/enhancements/issues/1965

### CODE REVIEW

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/highly-available-master.md:182]: I'd write this as:
```suggestion
The API Server Identity feature is controlled by a
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
and is not enabled by default. You can activate API Server Identity by enabling
the feature gate named `APIServerIdentity` when you start the
{{< glossary_tooltip text="API Server" term_id="kube-apiserver" >}}:
```

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/highly-available-master.md:190]:
```suggestion
During bootstrap, NAME_OF_THING assigns a unique ID to each API server. Each kube-apiserver
manages a [Lease](/docs/reference/generated/kubernetes-api/{{< param "version" >}}//#lease-v1-coordination-k8s-io) 
in the _kube-apiserver-lease_ {{< glossary_tooltip text="namespaces" term_id="namespace">}}.
The Lease contains the unique ID for the kube-apiserver. If another kube-apiserver tries to join
same control plane, the additional API server [picks a unique ID that is not used within the cluster?]

Enabling this feature [provides benefits?].

Each API server also exposes its own ID as a label dimension in its metrics: `TBD`.
```

**Review by sftim — COMMENTED** [review]
Hi. Here's some early feedback - I hope it's helpful.

**Comment by annajung** [comment]
/hold cancel
k/k prs merged

hi @roycaihw with k/k prs merged in, Sig Docs would appreciate it if you could get the docs content ready to be reviewed early, if possible, before kubecon. It will give us more time to make sure we get both tech/docs review on time before the release. Thank you! 

Please don't forget to change the PR title once it's ready to be reviewed. (remove wip)

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (annajung) [inline content/en/docs/tasks/administer-cluster/highly-available-master.md:181]: Hi there, please don't forget to add `APIServerIdentity` to https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/ and a short description of what functionality it provides. thank you!

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (sftim) [reply content/en/docs/tasks/administer-cluster/highly-available-master.md:182]: or:
```suggestion
_API Server identity_ is behind a
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
named `APIServerIdentity`, that is not enabled by default.
You can enable API Server Identity by adding the following command-line
flag to your `kube-apiserver` invocation:
```

File: content/en/docs/concepts/overview/working-with-objects/namespaces.md
Comment (sftim) [inline content/en/docs/concepts/overview/working-with-objects/namespaces.md:66]: Remember to change “four” to “several”. Also, I suggest sorting the list.

Here are some wording suggestions that can either go in this PR _or_ land after this one merges.

```suggestion
   * `kube-apiserver-lease` *⚠????TODO????TODO????⚠* _behind feature gate_
   * `kube-node-lease` A namespace for the lease objects associated with each node. Node leases improve the performance of node heartbeats for large clusters.
   * `kube-public` A namespace that is created automatically and is readable by all users (including those not authenticated). This namespace is mostly reserved for cluster usage, in case some resources should be visible and readable publicly throughout the whole cluster. 
   * `kube-system` General namespace for Kubernetes system objects.

{{< note >}}
Making the `kube-public` namespace and its contents world-readable is a convention, not a requirement.
{{< /note >}}
```

**Comment by reylejano-rxm** [comment]
Hi @roycaihw , please change the PR title to remove the words "wip: placeholder" when the PR is ready.

**Comment by reylejano-rxm** [comment]
@roycaihw, please keep in mind that the Docs PR Ready for Review deadline is coming up on Monday, November 23

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (roycaihw) [reply content/en/docs/tasks/administer-cluster/highly-available-master.md]: Thanks. Updated the paragraph with the first suggestion.

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (roycaihw) [reply content/en/docs/tasks/administer-cluster/highly-available-master.md]: Updated the mechanism and the benefit. 

The metrics didn't made it into alpha. Dropped the sentence.

File: content/en/docs/concepts/overview/working-with-objects/namespaces.md
Comment (roycaihw) [reply content/en/docs/concepts/overview/working-with-objects/namespaces.md]: Thanks for the suggestion. We decided to reuse `kube-system` for this feature. Reverted the change.

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (roycaihw) [reply content/en/docs/tasks/administer-cluster/highly-available-master.md]: Done. Copied the one-sentence descriptions from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/features/kube_features.go

**Comment by roycaihw** [comment]
It's ready for review. Please take a look. Thanks!

**Comment by reylejano-rxm** [comment]
@kubernetes/sig-api-machinery-pr-reviewer , can we get a tech review (lgtm)
Thank you

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:154]: Is this intended? The PR title doesn't mention `StorageVersionAPI`.

If I were documenting 2 different features I'd probably do that in 2 separate commits - it's easier to revert if ever needed.

**Review by sftim — CHANGES_REQUESTED** [review]
I'm puzzled about the StorageVersionAPI feature gate that this PR is now adding.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md
Comment (lavalamp) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:154]: Yes there's two separate things, storage version depends on the identity flag being on. @roycaihw is there a second PR documenting storage version? Should these be moved there?

Comment (roycaihw) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md]: Storage version API is not user-facing. The only user of it is the [storage migrator](https://github.com/kubernetes-sigs/kube-storage-version-migrator). Currently there is no plan to have a k8s.io doc for it. We may document it in the storage migrator repo. cc @caesarxuchao 

I removed this line. 

(we have the k8s.io doc for apiserver-identity because it can be used for other features, e.g. priority & fairness)

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (caesarxuchao) [inline content/en/docs/tasks/administer-cluster/highly-available-master.md:195]: Shall we also introduce the lease duration and gc period flags?

**Review by caesarxuchao — COMMENTED** [review]
LGTM otherwise.

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/highly-available-master.md:198]: nits (we avoid Latin abbreviations; typo fix)
```suggestion
Enabling this feature is a prerequisite for using features that involve HA API
server coordination (for example, the StorageVersion API feature).
```

Is StorageVersion a `kind` or the name of a feature gate? I'm not sure readers can work that out.

**Review by sftim — COMMENTED** [review]
Hi, a bit more feedback.

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (roycaihw) [reply content/en/docs/tasks/administer-cluster/highly-available-master.md]: added the flags

**Comment by annajung** [comment]
Hi @roycaihw if you address some of the comments above, I think this will be able to merge soon.

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (roycaihw) [reply content/en/docs/tasks/administer-cluster/highly-available-master.md]: updated with details. Thanks

File: content/en/docs/tasks/administer-cluster/highly-available-master.md
Comment (roycaihw) [reply content/en/docs/tasks/administer-cluster/highly-available-master.md]: it's the feature gate. Updated

**Comment by annajung** [comment]
Page preview: https://deploy-preview-24921--kubernetes-io-vnext-staging.netlify.app/docs/tasks/administer-cluster/highly-available-master/#api-server-identity

/lgtm

**Pull request merged**
